### PR TITLE
refactor: Tidy up builders and components

### LINKED ISSUE #7702: ModalBuilder does not accept camel cases
### Which package is this bug report for?

builders

### Issue description

Hi, I've heard that builders are expected to accept camel cases,
but I've noticed that ModalBuilder does not accept camel cases, while others(like TextInputBuilder) do.

### Code sample

```typescript
new ModalBuilder({
    customId: 'something' // will not work
    custom_id: 'something' // will work
})
```

### Package version

14.0.0-dev.1647259751.2297c2b

### Node.js version

16.14.0

### Operating system

_No response_

### Priority this issue should have

Low (slightly annoying)

### Which partials do you have configured?

Not applicable (subpackage bug)

### Which gateway intents are you subscribing to?

Not applicable (subpackage bug)

### I have tested this issue on a development release

72577c4

### PR DESCRIPTION
**Pull request by Jiralite**
**Please describe the changes this PR makes and why it should be merged:**
Some builders and components were not documented (and one was not added). This pull request adds them all in. Here is a summary:

- Documented various builders and components
- Corrected 'json' capitalisation
- Fixed `@external` references. They were being overwritten by discord.js's classes and thus did not work
- Added valid documentation to `TextInputComponent` which would otherwise cause the documentation generator to fail
- Used consistent renaming for importing builders
- Fixes an amazing typo that caused `MessageComponentData` to not appear in the documentation
- Resolves #7702 by adding the missing `ModalBuilder` to discord.js

**Status and versioning classification:**
- This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

**Comment by almeidx** [comment]
Can you also add an extended ModalBuilder in djs to support camel cased properties?

File: packages/discord.js/src/index.js [hunk 1]
```diff
@@ -208,6 +208,8 @@ exports.TextInputStyle = require('discord-api-types/v10').TextInputStyle;
 exports.UserFlags = require('discord-api-types/v10').UserFlags;
 exports.WebhookType = require('discord-api-types/v10').WebhookType;
 exports.UnsafeButtonBuilder = require('@discordjs/builders').UnsafeButtonBuilder;
+exports.UnsafeEmbed = require('@discordjs/builders').UnsafeEmbedBuilder;
+exports.UnsafeModal = require('@discordjs/builders').UnsafeModalBuilder;
```
Comment (muchnameless) [inline packages/discord.js/src/index.js:212]:
```suggestion
exports.UnsafeEmbedBuilder = require('@discordjs/builders').UnsafeEmbedBuilder;
exports.UnsafeModalBuilder = require('@discordjs/builders').UnsafeModalBuilder;
```

File: packages/discord.js/src/structures/ModalBuilder.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+const { ModalBuilder: BuildersModal, isJSONEncodable } = require('@discordjs/builders');
+const Transformers = require('../util/Transformers');
+
+/**
+ * Represents a modal builder.
+ */
+class ModalBuilder extends BuildersModal {
+  constructor({ components, ...data }) {
+    super({
+      components: components?.map(c => (isJSONEncodable(c) ? c.toJSON() : Transformers.toSnakeCase(c))),
```
Comment (muchnameless) [inline packages/discord.js/src/structures/ModalBuilder.js:12]: shouldn't this do an `instanceof Component` check instead like `ActionRowBuilder`'s constructor?

Comment (suneettipirneni) [reply packages/discord.js/src/structures/ModalBuilder.js:12]: no, I recommended in discord to use `isJSONEncodable`, action row should do the same at some point.

Comment (Jiralite) [reply packages/discord.js/src/structures/ModalBuilder.js]: We decided to go with the `instanceof` operator as to not lose references to component classes.

File: packages/discord.js/typings/index.d.ts [hunk 3]
```diff
@@ -559,6 +560,11 @@ export class SelectMenuOptionBuilder extends BuildersSelectMenuOption {
   public setEmoji(emoji: ComponentEmojiResolvable): this;
 }
 
+export class ModalBuilder extends BuildersModal {
+  public constructor(data?: ModalData | APIModalComponent);
```
Comment (suneettipirneni) [inline packages/discord.js/typings/index.d.ts:564]: The `ModalData` interface needs to be updated to allow both json components and class components

**Review by suneettipirneni — CHANGES_REQUESTED** [review]

**Review by kyranet — APPROVED** [review]

File: packages/discord.js/src/structures/ModalBuilder.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+const Transformers = require('../util/Transformers');
+
+/**
+ * Represents a modal builder.
+ */
+class ModalBuilder extends BuildersModal {
+  constructor({ components, ...data }) {
+    super({
+      components: components?.map(c => (c instanceof ComponentBuilder ? c : Transformers.toSnakeCase(c))),
+      ...Transformers.toSnakeCase(data),
```
Comment (vladfrangu) [inline packages/discord.js/src/structures/ModalBuilder.js:13]: Invert these

**Review by vladfrangu — CHANGES_REQUESTED** [review]

**Review by vladfrangu — APPROVED** [review]

**Review by muchnameless — COMMENTED** [review]
```ts
  /**
   * Creates a new action row builder from json data
   * @param {JSONEncodable<APIActionRowComponent> | APIActionRowComponent} other The other data
   * @returns {ActionRowBuilder}
   */
  static from(other) {
    if (isJSONEncodable(other)) {
      return new this(other.toJSON());
    }
    return new this(other);
  }
```
this is currently the only re-exported builder without a `from` method

**Review by suneettipirneni — APPROVED** [review]

**Review by SpaceEEC — APPROVED** [review]

**Comment by iCrawl** [comment]
This needs a rebase.

**Review by suneettipirneni — CHANGES_REQUESTED** [review]
I forgot about this, but can you also add an external link to JSONEncodable https://discord.js.org/#/docs/builders/main/typedef/JSONEncodable

**Review by suneettipirneni — APPROVED** [review]

**Pull request merged**